Implement MVP Feedbot

dbp opened the pull request:
This adds support for autograders providing custom LLM-backed feedback to parts of assignments. 

It is both minimal -- it requires no database change, and minimal UI changes -- and flexibly autograder-agnostic: any autograder can easily plug into the mechanism, using it for individual problems, for entire assignments, or for really anything where executing a prompt on-demand can be useful for student. It's also essentially a direct port of the existing https://github.com/NUFeedBot/feedbot-server architecture -- and takes from that the one-shot nature -- there is no ability to get _additional_ rounds of feedback, or refinement on the results, though that's certainly something we could explore in the future. There also isn't currently any rate limited implemented -- while there is natural rate limiting based on the fact that everything is on-demand, it may be desirable to put overall caps per student, or per assignment, and display an error if they are hit -- this would fit in pretty well with the existing architecture.

To use it, autograder should add:
```
...
"extra_data": { 
  "llm": { 
    "prompt": "... full prompt here, including code if needed ...",
    "model": "...",
    "provider": "...", // "azure" or "openai" or "anthropic"
    "account" : "...", // string that selects env var; if missing, uses default
    "temperature": ..., // optional
    "max_tokens": ..., // optional
    "type": "v1"
  }
 ...
}
...
```

To a test result to turn it into an LLM test result. This changes the UI to:

1. Use a robot icon for test cases.
2. Replace the message for the test case with, initially, a button to generate the feedback:

<img width="1257" height="373" alt="Screenshot 2025-08-30 at 1 47 43 PM" src="https://github.com/user-attachments/assets/2b405543-44ae-4400-a503-7ea587c55031" />

Once you click the button, an api call is made to openai, which sends the full prompt from `llm_hint_prompt` (exactly as written, no additions or changes), and the response is put in.

<img width="739" height="188" alt="Screenshot 2025-08-30 at 1 49 07 PM" src="https://github.com/user-attachments/assets/dff0d59a-36b4-4477-80d1-222f2ec19877" />

<img width="942" height="337" alt="Screenshot 2025-08-30 at 1 49 26 PM" src="https://github.com/user-attachments/assets/a6987c82-7247-425a-8318-654ad580baf3" />

Note, importantly, that the _prompt_ is never displayed to the user, though it is stored (permanently), in case it is helpful for debugging, research, etc. 

The response, in addition to being rendered, is stored in `llm.result` in `extra_data`, and on future page loads this is rendered rather than the button.

Note that while the intention is that test results that use this feature always have `score` and `max_score` of 0, since we intentionally did not alter any of the totalling logic, we still display the point values in both of the summary tables, in case the scores are _not_ 0/0. 

One final note -- currently this points against the OpenAI API -- we should change this to the AzureOpenAI API -- that's a couple line change though.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - AI-assisted hints for test results with caching, per-test feedback, usage tracking, provider support, and robot UI indicator.
  - Embedded interactive Pyret REPLs in results (student & instructor views).

- **Bug Fixes / Improvements**
  - Unified result formatting and icon/status logic that prioritizes AI hints; invitations query now omits accepted invites; improved retry/auth guards for integrations and live-meeting callbacks; minor comment fixes.

- **Database & Metrics**
  - New tables for hint feedback, LLM usage, and video meeting participation; expanded metrics RPCs and Grafana dashboard.

- **Documentation**
  - Local Supabase setup guide and detailed metrics documentation.

- **Tests**
  - New unit and end-to-end tests for LLM hints and Pyret REPL; Jest/test tooling added.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dbp committed:
implement MVP feedbot

jon-bell commented on app/api/llm-hint/route.ts:
```diff
@@ -0,0 +1,182 @@
+
+    if (updateError) {
+      console.error("Error storing LLM hint result:", updateError);
+      // Still return the result even if storage fails
```
throughout the API handler, please use the Sentry API to add tags/breadcrumbs and log errors

jon-bell reviewed (commented):
Looks like a great first pass! I think that collecting feedback on usefulness is important, even for MVP... here is a prompt for an LLM:

Add a new supabase migration to this PR, introducing a new table: assignment_hint_feedback, columns: class_id int8, submission_id int8, hint text, boolean useful, text comment, uuid created_by. class is FK to classes, submission FK to submissions, created_by FK to profiles. RLS on read is authorize_for_submission(submission_id), on insert is that PLUS authorizeforprofile(created_by). Add a simple react component that allows for submitting feedback after viewing a hint. Examine past migrations to ensure that the columns/types etc match up and use the same format.

dbp commented on app/api/llm-hint/route.ts:
```diff
@@ -0,0 +1,182 @@
+      });
+    }
+
+    // Call OpenAI API with the prompt from the database
+    const completion = await openai.chat.completions.create({
+      model: process.env.OPENAI_MODEL || "gpt-4o-mini",
+      messages: [
+        {
+          role: "user",
+          content: extraData.llm_hint_prompt
+        }
+      ],
+      max_tokens: 1000,
+      temperature: 0.7
+    });
+
```
valid for attribution but deprecated. Ho hum.

dbp committed:
feedbot: refactor extra data to llm subobject, make more configurable; other papercuts

dbp committed:
feedbot: add feedback

dbp commented:
> Looks like a great first pass! I think that collecting feedback on usefulness is important, even for MVP... here is a prompt for an LLM:
> 
> Add a new supabase migration to this PR, introducing a new table: assignment_hint_feedback, columns: class_id int8, submission_id int8, hint text, boolean useful, text comment, uuid created_by. class is FK to classes, submission FK to submissions, created_by FK to profiles. RLS on read is authorize_for_submission(submission_id), on insert is that PLUS authorizeforprofile(created_by). Add a simple react component that allows for submitting feedback after viewing a hint. Examine past migrations to ensure that the columns/types etc match up and use the same format.

I did this, with some refinement. It, along with other changes, reflected in current state of this branch.

jon-bell committed:
Fix types, add more configurability, clean up UI to work with dark mode, add a test, add usage logging

ironm00n committed:
Embedded Pyret REPL (#170)

New Features

Embedded interactive Pyret REPL in grader results with expandable panels and instructor-only mode.
UI/UX

Loading spinner, clear error states, collapsible behavior, and “No output” fallback when results are empty.
Tests

Comprehensive end-to-end suite covering student/instructor flows, multi-REPL scenarios, error handling, and visual snapshots.
Chores

Added REPL runtime dependency; tooling/dev config updates (local dev shells, env handling, gitignore, .envrc); admin client and test-utils adjustments; small CLI/script improvements; added assert helper.

Documentation

Expanded README with local Supabase setup instructions.

jon-bell committed:
More metrics w/ grafana dashboard, improve SIS reliability, receive events from chime SDK

New Features
Auto-accepts matching SIS invitations and backfills pending invites.
Adds video meeting participant tracking and richer video-meeting metrics.
Expands Prometheus metrics and adds a Grafana dashboard plus metrics documentation.
Bug Fixes
Hides already-accepted invitations from enrollment views and related exports.
Chores
Improves SIS API reliability with retries and reduces noisy error reporting.

dbp commented on app/api/llm-hint/route.ts:
```diff
@@ -1,31 +1,100 @@
+  temperature?: number;
+  maxTokens?: number;
+  maxRetries?: number;
+  account?: string;
```
Is the account just used for testing? i.e., switching between API keys?

jon-bell replied:
for co-tenancy on pawtograder, where different classes have different inference API keys

jon-bell committed:
Update seed script, streamline feedback

dbp committed:
Merge remote-tracking branch 'origin/staging' into feedbot

dbp committed:
oops, missing file

dbp committed:
Add configurable feedbot rate limiting

jon-bell committed:
unit test

jon-bell committed:
linting, remove 'any' types

jon-bell committed:
also fix bad merge :(

jon-bell committed:
fix test setup, add metrics

dbp committed:
fix for when model responds with nothing, due to running out of tokens

dbp committed:
change llm output type to still show output, above form; that way output can be used to explain feedbot

jon-bell committed:
final metrics edits

jon-bell committed:
Merge branch 'feedbot' of github.com:pawtograder/platform into feedbot

jon-bell committed:
sneak in an invite fix, last coderabbit stuff
